### ISSUE TITLE
[Web] Different model output with WebGPU vs WASM (or Python with the CUDA EP)

### ISSUE THREAD

**Issue by Dexterp37** [microsoft/onnxruntime#24070]
### Describe the issue

Using the model referenced below (see STR) produces a substantially different output when using the WebGPU provider (**wrong**) vs the WASM provider (**correct**). The difference is big enough in the output to indicate some bug in some WebGPU operator.

To further confirm this, I ran inference with the same model using onnx in Python: the correct results are consistent with both the CUDA and the CPU providers.

I'm lost at how to debug this further unfortunately.

### To reproduce

1. Host the following web page with a minimal example. Note that it's using 1.21.0 from a CDN, but the result doesn't change even with the latest available dev version. The needed files can be downloaded: [model](https://drive.google.com/file/d/1Zyv1iDGFEX3DqgJ5MtuIPZDf6YFaLVC3/view?usp=sharing) and [input image](https://drive.google.com/file/d/1RYjdtl2cCUFqzwsmJWKcNWKkNaN9HH0b/view?usp=sharing).

```html
<html>

<head>
  <script src="https://cdn.jsdelivr.net/npm/onnxruntime-web@1.21.0/dist/ort.min.js"></script>
</head>

<body>
    <h1>Example</h1>
    <div>
        <img id="image" src="turtle3.png" alt="Image preview..." width="1333" height="750" hidden>
        <canvas id="mycanvas" width="1333" height="750"></canvas>
    </div>

</body>
<script>
    async function startInferenceSession(modelPath) {
      const session = await ort
        .InferenceSession
        .create(
          modelPath,
          {
            executionProviders: [
              // 'wasm',
              'webgpu'
            ],
            graphOptimizationLevel: 'all',
          }
        );
      console.log('Inference session created');
      return session;
    }


    async function runInference(session, input) {
      const feeds = {};
      feeds[session.inputNames[0]] = input;
      const outputData = await session.run(feeds);
      return outputData;
    }
    
    async function imageToTensor(
      imageData,
    ) {
      const mean = [0.485, 0.456, 0.406]
      const std = [0.229, 0.224, 0.225]

      const rawTensor = await ort.Tensor.fromImage(imageData)

      // Normalize the tensor as needed.
      const rawData = rawTensor.data
      const [_, C, H, W] = rawTensor.dims;

      const normalizedData = new Float32Array(rawData.length)

      const dataPerChannel = H * W;
      for (let c = 0; c < C; c++) {
        for (let h = 0; h < H; h++) {
          for (let w = 0; w < W; w++) {
            const i = c * dataPerChannel + h * W + w;
            normalizedData[i] = (rawData[i] - mean[c]) / std[c];
          }
        }
      }
      return new ort.Tensor(normalizedData, rawTensor.dims);
    }

    async function testInference(session) {
      console.log('Testing inference');
      let canvas = document.getElementById("mycanvas");
      let ctx = canvas.getContext("2d");
      const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height)

      const tensor = await imageToTensor(imageData)
      console.log('Tensor:', tensor)

      const outData = await runInference(session, tensor)

      const { pred_boxes, logits } = outData
      console.log('logits (first 10 values):', logits.data.slice(0, 10))
    }
  
    async function main() {
        let img = document.getElementById('image');
        let canvas = document.getElementById("mycanvas");
        var ctx = canvas.getContext("2d");
        ctx.drawImage(img, 0, 0, 1333, 750);

        let session = await startInferenceSession("./test.onnx");
        await testInference(session);
    }
    main();

</script>
</html>

```

2. With the wasm provider, the first logit values (see the console logs) are: `[-9.01431655883789, -9.044118881225586, -6.852304458618164, -8.999532699584961, -9.034984588623047, -8.848183631896973, -8.849029541015625, -6.6562042236328125, -8.858226776123047, -8.890541076660156]`. These numbers are correct and match my ground truth.
3. With webgpu the first logit values (see the console logs) are  `[-9.747684478759766, -9.810420989990234, -8.110831260681152, -9.75857162475586, -9.758389472961426, -9.946432113647461, -9.912216186523438, -8.000043869018555, -9.904242515563965, -9.92009449005127]`. These numbers are quite a bit off.

### Urgency

This issue is urgent and it's blocking my research, unfortunately.

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.22.0-dev.20250310-fe7634eb6f

### Execution Provider

'webgpu' (WebGPU)

**Comment by xadupre** [comment]
+ @guschmue

**Comment by prathikr** [comment]
Issue appears to be with GridSample operator, will investigate further.

**Comment by prathikr** [comment]
@Dexterp37 please retry with latest dev build

**Comment by fs-eire** [comment]
> [@Dexterp37](https://github.com/Dexterp37) please retry with latest dev build

In case you don't want to build, you can use the artifacts from the CI pipeline here: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1653422&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by Dexterp37** [comment]
Thank you both @prathikr and @fs-eire . I just tested the version built from the CI and values from WebGPU are now much closer to the ones from Wasm: `[-9.014312744140625, -9.044116020202637, -6.852304935455322, -8.999527931213379, -9.034981727600098, -8.848180770874023, -8.849030494689941, -6.6562066078186035, -8.858227729797363, -8.890539169311523]` . They are identical up to the 4th or 5th decimal place.

However, part of the model output is still off. Adding:

>      console.log('pred_boxes (first 10 values):', pred_boxes.data.slice(0, 10))

Right below the logits print, highlights a new issue:

- With the WASM version, the first 10 values are `[0.8819793462753296, 0.33753177523612976, 0.19872236251831055, 0.062045931816101074, 0.3585531711578369, 0.7631466388702393, 0.4469307065010071, 0.1789032518863678, 0.6021042466163635, 0.961134672164917]`
- With the WebGPU version, the first 10 values are  `[0.8819794058799744, 0.6150027513504028, 0.1987222135066986, 0.06204574182629585, 0.49718883633613586, 0.690393328666687, 0.44693082571029663, 0.17890317738056183, 0.5091800689697266, 0.845749020576477]`

Note how the second value is quite different in both cases.

**Comment by prathikr** [comment]
Issue is with ScatterND kernel, will take a look

**Comment by prathikr** [comment]
@Dexterp37 thank you for your patience, issue should be resolved now. Please retry and confirm.

**Comment by Dexterp37** [comment]
@prathikr thank you, I'll check this in the next couple of days.

**Comment by fs-eire** [comment]
The artifacts get expired in the link above. You can use https://github.com/microsoft/onnxruntime/actions/runs/14149066733 to download the artifact

**Comment by Dexterp37** [comment]
@prathikr @fs-eire thanks for your patience on this, and your work, apologies if it took me a while to get back to you but I had to verify what was happening. **tl;dr - we're getting closer, but I believe there's some other bug somewhere else!**

The values the I outlined in my previous comments are now matching, up to the 5th decimal numbers (there's one weird outlier: a number that matches exactly, `0.7631466388702393`). For example:

* With WASM: `[0.8819793462753296, 0.33753177523612976, 0.19872236251831055, 0.062045931816101074, 0.3585531711578369, 0.7631466388702393, 0.4469307065010071, 0.1789032518863678, 0.6021042466163635, 0.961134672164917]`
* With WebGPU: `[0.8819794058799744, 0.33753278851509094, 0.1987222135066986, 0.06204574182629585, 0.35855308175086975, 0.7631466388702393, 0.44693082571029663, 0.17890317738056183, 0.6021043658256531, 0.9611347317695618]`

I was tempted to discount this as a numerical approximation difference between the two EP, but maybe it's not? This (or something else!) seem to still have impact on the detections.

To make this easy to reproduce, I slightly modified once more the example I provided above to directly look up the offending box/data. I know the index of the best bounding box and the best score in the tensors:

```typescript
    async function testInference(session) {
      console.log('Testing inference');
      let canvas = document.getElementById("mycanvas");
      let ctx = canvas.getContext("2d");
      const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height)

      const tensor = await imageToTensor(imageData)
      console.log('Tensor:', tensor)

      const outData = await runInference(session, tensor)

      const { pred_boxes, logits } = outData
      console.log('logits (first 10 values):', logits.data.slice(0, 10))
      console.log('pred_boxes (first 10 values):', pred_boxes.data.slice(0, 10))

      const sigmoid = (x) => 1 / (1 + Math.exp(-x))
      const prob = logits.data.map(sigmoid)
      const probTensor = new ort.Tensor("float32", prob, logits.dims)
      console.debug("Best score (known)", probTensor.data[617])
      console.debug("Best box raw (known)", pred_boxes.data.slice(123 * 4, (123 + 1) * 4))
    }
```

The outputs:

* With WASM:

```
Best score (known) 0.9043878316879272
Best box raw (known) Float32Array(4) [0.5065099596977234, 0.48670125007629395, 0.03329381346702576, 0.05971693992614746]
```

* With WebGPU: 
```
Best score (known) 0.9043883681297302
Best box raw (known) Float32Array(4) [0.490261971950531, 0.4262031018733978, 0.03329379856586456, 0.05971701070666313]
```

Note how the coordinates of the bounding box differ on the first decimal already.

**Comment by prathikr** [comment]
Hi @Dexterp37 we've done more analysis and determined that you're running into an edge case with ScatterND where reduction is set to none but there are some indices with multiple updates. According to the [onnx spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#ScatterND) this is undefined behavior. We've tried a few different methods to try and account for this within the shader itself but it is proving difficult in your specific case because the indices tensor is very large.

Is it possible for you to share how you generated the onnx file for your model? We can loop in some exporter/converter folks to maybe attack the problem from there so you do not run into an undefined behavior case.

**Comment by prathikr** [comment]
@Dexterp37 thanks for the quick reply. Apologies for the confusion, I meant could I have access to the model code so I can see which line is generating the scatter-nd kernel and potentially adjust that.

**Comment by Dexterp37** [comment]
> [@Dexterp37](https://github.com/Dexterp37) thanks for the quick reply. Apologies for the confusion, I meant could I have access to the model code so I can see which line is generating the scatter-nd kernel and potentially adjust that.

Ah apologies! The model code comes from HF transformer's library. I disabled [custom kernels](https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/models/deformable_detr/configuration_deformable_detr.py#L127) to allow exporting to onnx.

I'm happy to share the non onnx model privately if that helps

**Comment by prathikr** [comment]
Awesome, can you email model files and exporting logic to prathikrao@microsoft.com

**Comment by prathikr** [comment]
@Dexterp37 we are in the process of migrating from javascript execution provider implementing webgpu kernels to a native webgpu execution provider. I tested your model with native and all outputs (logits, pred_boxes, best_score, and best_box_raw) are correct. Please retry your original script with the following import to run your model with webgpu native:

```Javascript
import * as ort from 'https://cdn.jsdelivr.net/npm/onnxruntime-web@dev/dist/ort.webgpu.bundle.min.mjs';
```

**Comment by prathikr** [comment]
Hi @Dexterp37 we made some fixes, please try with the following import (note the onnxruntime-web@dev):

```html
<script src="https://cdn.jsdelivr.net/npm/onnxruntime-web@dev/dist/ort.min.js"></script>
```

**Comment by satyajandhyala** [comment]
@Dexterp37 Please try downloading the latest dev build from [this link](https://artprodcus3.artifacts.visualstudio.com/Abc038106-a83b-4dab-9dd3-5a41bc58f34c/530acbc4-21bc-487d-8cd8-348ff451d2ff/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2FpaW5mcmEvcHJvamVjdElkLzUzMGFjYmM0LTIxYmMtNDg3ZC04Y2Q4LTM0OGZmNDUxZDJmZi9idWlsZElkLzc4MzYyNC9hcnRpZmFjdE5hbWUvb25ueHJ1bnRpbWUtd2Vi0/content?format=file&subPath=%2Fonnxruntime-web-1.23.0-dev.20250509-3dc91e6c31.tgz)

**Comment by satyajandhyala** [comment]
The above link downloads the build artifact. Alternatly you could install using `npm install onnxruntime-web@1.23.0-dev.20250509-3dc91e6c31`

**Comment by Dexterp37** [comment]
> [@Dexterp37](https://github.com/Dexterp37) Please try downloading the latest onnxruntime-web dev build from [this link](https://artprodcus3.artifacts.visualstudio.com/Abc038106-a83b-4dab-9dd3-5a41bc58f34c/530acbc4-21bc-487d-8cd8-348ff451d2ff/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2FpaW5mcmEvcHJvamVjdElkLzUzMGFjYmM0LTIxYmMtNDg3ZC04Y2Q4LTM0OGZmNDUxZDJmZi9idWlsZElkLzc4MzYyNC9hcnRpZmFjdE5hbWUvb25ueHJ1bnRpbWUtd2Vi0/content?format=file&subPath=%2Fonnxruntime-web-1.23.0-dev.20250509-3dc91e6c31.tgz)

This is awesome! This is now correctly working, the output is consistent with the output of WASM.

Great job folks, thank you @satyajandhyala and @prathikr !
